frame/scheduler/src/lib.rs

```diff
@@ -122,10 +148,57 @@ impl Default for Releases {
 	}
 }
 
+#[cfg(feature = "runtime-benchmarks")]
+mod preimage_provider {
+	use frame_support::traits::PreimageRecipient;
+	pub trait PreimageProviderAndMaybeRecipient<H>: PreimageRecipient<H> {}
+	impl<H, T: PreimageRecipient<H>> PreimageProviderAndMaybeRecipient<H> for T {}
+}
+
+#[cfg(not(feature = "runtime-benchmarks"))]
+mod preimage_provider {
+	use frame_support::traits::PreimageProvider;
+	pub trait PreimageProviderAndMaybeRecipient<H>: PreimageProvider<H> {}
+	impl<H, T: PreimageProvider<H>> PreimageProviderAndMaybeRecipient<H> for T {}
+}
+
+pub use preimage_provider::PreimageProviderAndMaybeRecipient;
+
+pub(crate) trait MarginalWeightInfo: WeightInfo {
+	fn item(periodic: bool, named: bool, resolved: Option<bool>) -> Weight {
+		match (periodic, named, resolved) {
+			(_, false, None) => Self::on_initialize_aborted(2) - Self::on_initialize_aborted(1),
+			(_, true, None) =>
+				Self::on_initialize_named_aborted(2) - Self::on_initialize_named_aborted(1),
+			(false, false, Some(false)) => Self::on_initialize(2) - Self::on_initialize(1),
+			(false, true, Some(false)) =>
+				Self::on_initialize_named(2) - Self::on_initialize_named(1),
+			(true, false, Some(false)) =>
+				Self::on_initialize_periodic(2) - Self::on_initialize_periodic(1),
+			(true, true, Some(false)) =>
+				Self::on_initialize_periodic_named(2) - Self::on_initialize_periodic_named(1),
+			(false, false, Some(true)) =>
+				Self::on_initialize_resolved(2) - Self::on_initialize_resolved(1),
+			(false, true, Some(true)) =>
+				Self::on_initialize_named_resolved(2) - Self::on_initialize_named_resolved(1),
+			(true, false, Some(true)) =>
+				Self::on_initialize_periodic_resolved(2) - Self::on_initialize_periodic_resolved(1),
+			(true, true, Some(true)) =>
+				Self::on_initialize_periodic_named_resolved(2) -
+					Self::on_initialize_periodic_named_resolved(1),
```

Comment:
will come back to this one, but lots of unsafe math here. Assumption seems to be that `f(2) > f(1)`, but that may not be true.

Reply:
It seems this is only used in tests. Let's just move this there then?

Reply:
There is no way for a higher parameter to be less than a lower parameter since we (rightly) don't allow for negative coefficients.

The fact that this is *after* a linear regression is why it makes sense. This function basically just extracts the `m` component of the curve `y=mx+C`; this is the cleanest way of working with a loop in which each loop element may go down one of several branches.